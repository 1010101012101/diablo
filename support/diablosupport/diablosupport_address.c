/* This research is supported by the European Union Seventh Framework Programme (FP7/2007-2013), project ASPIRE (Advanced  Software Protection: Integration, Research, and Exploitation), under grant agreement no. 609734; on-line at https://aspire-fp7.eu/. */

/* The development of portions of the code contained in this file was sponsored by Samsung Electronics UK. */

#include <diablosupport.h>

t_string
IoModifierAddressGeneric (t_const_string modifiers, va_list * ap)
{
  char buffer[20];
  t_string ret;
  t_string_array *array;
  t_address_generic addr = va_arg (*ap, t_address_generic);

  array = StringArrayNew ();
  if (addr.type == ADDRSIZE32)
    sprintf (buffer, "0x%x", addr.types.addr32);
  else if (addr.type == ADDRSIZE64)
    sprintf (buffer, "0x%"PRIx64"", addr.types.addr64);
  else
    FATAL(("Address not correctly initialized!"));

  StringArrayAppendString (array, StringDup (buffer));

  ret = StringArrayJoin (array, " ");
  StringArrayFree (array);
  return ret;
}

t_string
IoModifierAddress32 (t_const_string modifiers, va_list * ap)
{
  char buffer[20];
  t_string ret;
  t_string_array *array;
  t_uint32 addr = va_arg (*ap, t_uint32);

  array = StringArrayNew ();
  sprintf (buffer, "0x%x", addr);

  StringArrayAppendString (array, StringDup (buffer));

  ret = StringArrayJoin (array, " ");
  StringArrayFree (array);
  return ret;
}

t_string
IoModifierAddress64 (t_const_string modifiers, va_list * ap)
{
  char buffer[20];
  t_string ret;
  t_string_array *array;
  t_uint64 addr = va_arg (*ap, t_uint64);

  array = StringArrayNew ();
  sprintf (buffer, "0x%"PRIx64"", addr);

  StringArrayAppendString (array, StringDup (buffer));

  ret = StringArrayJoin (array, " ");
  StringArrayFree (array);
  return ret;
}

t_address_generic
RealAddressNew32 (t_const_string file, t_uint32 line, t_uint32 a)
{
  t_address_generic ret;

  ret.type = ADDRSIZE32;
  ret.types.addr32 = a;
  return ret;
}

t_address_generic
RealAddressNew64 (t_const_string file, t_uint32 line, t_uint64 a)
{
  t_address_generic ret;

  ret.type = ADDRSIZE64;
  ret.types.addr64 = a;
  return ret;
}

t_address_generic
RealAddressAdd (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s %d %d", line,
          file, a.type, b.type));

  ret.type = b.type;

  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 + b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 + b.types.addr64;
  return ret;
}

t_address_generic
RealAddressAnd (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = b.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 & b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 & b.types.addr64;
  return ret;
}

t_address_generic
RealAddressOr (t_const_string file, t_uint32 line, t_address_generic a,
               t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = a.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 | b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 | b.types.addr64;
  return ret;
}

t_address_generic
RealAddressXor (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = a.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 ^ b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 ^ b.types.addr64;
  return ret;
}

t_address_generic
RealAddressNot (t_const_string file, t_uint32 line, t_address_generic a)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = ~a.types.addr32;
  else
    ret.types.addr64 = ~a.types.addr64;
  return ret;
}

t_address_generic
RealAddressShl (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = b.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 << b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 << b.types.addr64;
  return ret;
}

t_address_generic
RealAddressShr (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = b.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 >> b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 >> b.types.addr64;
  return ret;
}

t_address_generic
RealAddressAddInt32 (t_const_string file, t_uint32 line, t_address_generic a,
                     t_int32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressAddUint64 (t_const_string file, t_uint32 line, t_address_generic a,
                      t_uint64 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + (t_uint32)b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressAddUint32 (t_const_string file, t_uint32 line, t_address_generic a,
                      t_uint32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressInverseMaskUint32 (t_const_string file, t_uint32 line, t_address_generic a,
                              t_uint32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 & (~((t_uint32) (b)));
  else
    ret.types.addr64 = a.types.addr64 & (~((t_uint64) (b)));
  return ret;
}

t_address_generic
RealAddressAddInt64 (t_const_string file, t_uint32 line, t_address_generic a,
                     t_int64 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + (t_int32)b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressSubUint32 (t_const_string file, t_uint32 line, t_address_generic a,
                      t_uint32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 - b;
  else
    ret.types.addr64 = a.types.addr64 - b;
  return ret;
}

t_address_generic
RealAddressSubInt32 (t_const_string file, t_uint32 line, t_address_generic a,
                     t_int32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 - b;
  else
    ret.types.addr64 = a.types.addr64 - b;
  return ret;
}

t_address_generic
RealAddressSub (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Sub'ed different size addresses at line %d in %s", line, file));

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 - b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 - b.types.addr64;
  return ret;
}

t_bool
RealAddressIsNull (t_const_string file, t_uint32 line, t_address_generic a)
{
  if (a.type == ADDRSIZE32)
    return a.types.addr32 == 0;
  else
    return a.types.addr64 == ULL(0);
}

t_bool
RealAddressIsMax (t_const_string file, t_uint32 line, t_address_generic a)
{
  if (a.type == ADDRSIZE32)
    return a.types.addr32 == 0xFFFFFFFF;
  else
    return a.types.addr64 == ULL(0xFFFFFFFFFFFFFFFF);
}

t_bool
RealAddressIsGt (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));
  if (a.type == ADDRSIZE32)
    return a.types.addr32 > b.types.addr32;
  else
    return a.types.addr64 > b.types.addr64;
}

t_bool
RealAddressIsGe (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type), ("Added different size addresses"));
  if (a.type == ADDRSIZE32)
    return a.types.addr32 >= b.types.addr32;
  else
    return a.types.addr64 >= b.types.addr64;
}

t_bool
RealAddressIsLt (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
  {
    return a.types.addr32 < b.types.addr32;
  }
  else
  {
    return a.types.addr64 < b.types.addr64;
  }
}

t_bool
RealAddressIsLe (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type), ("Added different size addresses"));
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
  {
    return a.types.addr32 <= b.types.addr32;
  }
  else
  {
    return a.types.addr64 <= b.types.addr64;
  }
}

t_bool
RealAddressIsEq (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type),
         ("AddressEQ at line %d in %s has different type operands", line,
          file));
  if (a.type == ADDRSIZE32)
    return a.types.addr32 == b.types.addr32;
  else
    return a.types.addr64 == b.types.addr64;
}

void *
RealAddressAddDispl (t_const_string file, t_uint32 line, void *in,
                     t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return ((void *) (((char *) in) + addr.types.addr32));
  }
  else
  {
    return ((void *) (((char *) in) + addr.types.addr64));
  }
}

t_address_generic
RealAddressMulUint32 (t_const_string file, t_uint32 line, t_address_generic address,
                      t_uint32 div)
{
  t_address_generic ret;

  ret.type = address.type;
  if (address.type == ADDRSIZE32)
  {
    ret.types.addr32 = address.types.addr32 * div;
  }
  else if (address.type == ADDRSIZE64)
  {
    ret.types.addr64 = address.types.addr64 * ((t_uint64) div);
  }
  else
  {
    FATAL(("Address is corrupt (called at file %s, line %d) ! %d", file,
           line, address.type));
  }
  return ret;
}

t_address_generic
RealAddressDivUint32 (t_const_string file, t_uint32 line, t_address_generic address,
                      t_uint32 div)
{
  t_address_generic ret;

  ret.type = address.type;
  if (address.type == ADDRSIZE32)
  {
    ret.types.addr32 = address.types.addr32 / div;
  }
  else if (address.type == ADDRSIZE64)
  {
    ret.types.addr64 = address.types.addr64 / ((t_uint64) div);
  }
  else
  {
    FATAL(("Address is corrupt (called at file %s, line %d) ! %d", file,
           line, address.type));
  }
  return ret;
}

t_uint32
RealAddressExtractUint32 (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return addr.types.addr32;
  }
  if (addr.types.addr64 > 0xffffffffull)
  {
    /* check for overflow: assume there is no overflow if it is a negative
     * number that also fits in 32 bits */
    if ((addr.types.addr64 & 0xffffffff80000000ull) != 0xffffffff80000000ull)
      FATAL(("Overflow at line %d in file %s val = %"PRIx64"\n", line, file, addr.types.addr64));
  }
  return (t_uint32) addr.types.addr64;
}

t_uint64
RealAddressExtractUint64 (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return (t_uint64) addr.types.addr32;
  }
  return addr.types.addr64;
}

t_int32
RealAddressExtractInt32 (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return (t_int32) addr.types.addr32;
  }
  if (addr.types.addr64 > 0xffffffffull)
  {
    if ((addr.types.addr64 & 0xffffffff80000000ull) !=
        0xffffffff80000000ull)
      FATAL(("Overflow at line %d in file %s\n", line, file));
    else
      /* negative number */
      return (t_int32) ((t_int64) addr.types.addr64);
  }
  return (t_int32) ((t_uint32) addr.types.addr64);
}

t_int64
RealAddressExtractInt64SignExtend (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return (t_int32) addr.types.addr32;
  }
  else if (addr.type == ADDRSIZE64)
  {
    return (t_int64) (addr.types.addr64);
  }
  else
  {
    FATAL(("address is corrupt"));
  }
}

t_address_generic
RealAddressSignExtend (t_const_string file, t_uint32 line, t_address_generic x,
                       t_uint32 signbit)
{
  if (x.type == ADDRSIZE32)
  {
    x.types.addr32 = (x.types.addr32 ^ (1 << signbit)) - (1 << signbit);
  }
  else if (x.type == ADDRSIZE64)
  {
    x.types.addr64 = (x.types.addr64 ^ (1 << signbit)) - (1 << signbit);
  }
  else
  {
    FATAL(("address is corrupt"));
  }

  return x;
}

#ifdef _MSC_VER
inline t_address
AddressNewTyped(t_address d, t_uint32 type)
{ 
  t_address res;
  if (type == ADDRSIZE32)
    res = AddressNew32(d);
  else if (type == ADDRSIZE64)
    res = AddressNew64(d);
  else
    FATAL(("Unknown address type (%d)", type));
  return res;   
}

inline t_uint32
WordAlignedSize(t_uint32 s)
{
  t_uint32 res;
  if ((s) <= 8)
    res = 8;
  else if ((s) <= 16)
    res = 16;
  else if ((s) <= 32)
    res = 32;
  else if ((s) <= 64)
    res = 64;
  else
    FATAL(("Unsupported size: %d", s));
  return res;
}
#endif

t_uint32 AddressSizeInBytes(t_address_type address_type)
{
  return (address_type == ADDRSIZE32) ? 4 : 8;
}

/* vim: set shiftwidth=2 expandtab cinoptions=p5,t0,(0, foldmethod=marker tw=80 cindent: */
