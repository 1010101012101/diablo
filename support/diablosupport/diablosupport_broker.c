#include <diablosupport.h>
#include <stdarg.h>
#include <string.h>

static t_hash_table *brokerht = NULL;

typedef struct _t_broker_cbdef
{
  t_hash_table_node node;
  t_string_array *proto;
  void *impl;
  void *data;
  t_bool final;
  struct _t_broker_cbdef *next;
} t_broker_cbdef;

static t_broker_cbdef *
DiabloBrokerCallLookup (const char *name)
{
  if (!brokerht)
    return NULL;
  return HashTableLookup (brokerht, name);
}

t_bool
DiabloBrokerCallExists (const char *name)
{
  return (DiabloBrokerCallLookup (name) != NULL);
}

void
DiabloBrokerCallInstall (const char *name, const char *prototype, void *implementation, t_bool final, ...)
{
  va_list ap;
  t_string_array_elem *elem, *elem2;
  t_broker_cbdef *cb, *cb2;

  if (!brokerht)
    brokerht = HashTableNew (121, 0, (t_hash_func) StringHash, (t_hash_cmp) StringCmp, NULL);

  if ((cb = DiabloBrokerCallLookup (name)))
  {
    cb2 = Calloc (1, sizeof (t_broker_cbdef));
    cb2->proto = StringDivide (prototype, ",", FALSE, FALSE);
    cb2->data = NULL;
    cb2->impl = implementation;
    cb2->final = final;
    if (!final)
    {
      t_string_array *tmp;
      void *timpl;

      tmp = cb->proto;
      cb->proto = cb2->proto;
      cb2->proto = tmp;
      cb2->data = cb->data;
      cb2->final = cb->final;
      cb->final = final;
      cb->data = NULL;
      timpl = cb->impl;
      cb->impl = cb2->impl;
      cb2->impl = timpl;
      cb2->next = cb->next;
      cb->next = cb2;
    }
    else
    {
restart:
      elem2 = cb2->proto->first;
      STRING_ARRAY_FOREACH_ELEM(cb->proto, elem)
      {
        if (!(strncmp (elem->string, "const ", 6)))
        {
          if ((!(strncmp (elem2->string, "const ", 6))) || (!cb->next))
          {
            cb2->next = cb->next;
            cb->next = cb2;
            cb = cb2;
            break;
          }
          else
          {
            cb = cb->next;
            goto restart;
          }
        }
        else if (!(strncmp (elem2->string, "const ", 6)))
        {
          /* Call cb, first, then cb2 */
          t_string_array *tmp;
          void *timpl;

          tmp = cb->proto;
          cb->proto = cb2->proto;
          cb2->proto = tmp;
          cb2->data = cb->data;
          cb2->final = cb->final;
          cb->final = final;
          cb->data = NULL;
          timpl = cb->impl;
          cb->impl = cb2->impl;
          cb2->impl = timpl;
          cb2->next = cb->next;
          cb->next = cb2;
          break;
        }
        else if (strcmp (elem->string, elem2->string))
        {
          FATAL(("Prototype disambiguity for two broker calls: proto1 = %s, proto2=%s\n", StringArrayJoin (cb->proto, ","), prototype));
        }

        elem2 = elem2->next;
        if (!elem2)
          FATAL(("Cannot distiguish between two broker calls: proto1 = %s, proto2=%s\n", StringArrayJoin (cb->proto, ","), prototype));
      }
    }
  }
  else
  {
    cb = Calloc (1, sizeof (t_broker_cbdef));
    cb->proto = StringDivide (prototype, ",", FALSE, FALSE);
    cb->data = NULL;
    cb->impl = implementation;
    cb->final = final;
    HASH_TABLE_NODE_SET_KEY(&cb->node, StringDup (name));
    HashTableInsert (brokerht, cb);
  }

  STRING_ARRAY_FOREACH_ELEM(cb->proto, elem)
  {
    if (!(strncmp (elem->string, "const ", 6)))
    {
      if (cb->data)
        FATAL(("DiabloBrokerCallInstall: More then one const argument!"));
      else
      {
        va_start (ap, final);
        cb->data = va_arg (ap, void *);
      }
    }
  }

  if (cb->proto->nstrings > 10)
    FATAL(("To many arguments for broker call"));
}

t_bool
DiabloBrokerCall (const char *name, ...)
{
  t_bool execed = FALSE;
  va_list ap;
  int tel = 0;
  void *args[10];
  t_string_array_elem *elem;

  /* Lookup name in hash table */
  t_broker_cbdef *callback_definition = DiabloBrokerCallLookup (name);

  do
  {
    t_bool found = FALSE;

    while (callback_definition && (!found))
    {
      found = TRUE;
      va_start (ap, name);
      tel = 0;
      STRING_ARRAY_FOREACH_ELEM(callback_definition->proto, elem)
      {
        args[tel] = (void *) va_arg (ap, void *);

        if (!(strncmp (elem->string, "const ", 6)))
        {
          if (args[tel] != callback_definition->data)
          {
            found = FALSE;
            break;
          }
        }
        tel++;
      }
      if (!found)
        callback_definition = callback_definition->next;
    }

    if (!callback_definition)
      return execed;

    switch (callback_definition->proto->nstrings)
    {
      case 0:
        ((void (*)()) (callback_definition->impl)) ();
        break;
      case 1:
        ((void (*)(void *)) (callback_definition->impl)) (args[0]);
        break;
      case 2:
        ((void (*)(void *, void *)) (callback_definition->impl)) (args[0], args[1]);
        break;
      case 3:
        ((void (*)(void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2]);
        break;
      case 4:
        ((void (*)(void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3]);
        break;
      case 5:
        ((void (*)(void *, void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3], args[4]);
        break;
      case 6:
        ((void (*)(void *, void *, void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3], args[4], args[5]);
        break;
      case 7:
        ((void (*)(void *, void *, void *, void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3], args[4], args[5], args[6]);
        break;
      case 8:
        ((void (*)(void *, void *, void *, void *, void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7]);
        break;
      case 9:
        ((void (*)(void *, void *, void *, void *, void *, void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], args[8]);
        break;
      case 10:
        ((void (*)(void *, void *, void *, void *, void *, void *, void *, void *, void *, void *)) (callback_definition->impl)) (args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], args[8], args[9]);
        break;
    }

    execed = TRUE;
  }
  while ((!callback_definition->final) && (callback_definition = callback_definition->next));

  return execed;
}

static void
BrokerCbdefFree (void *hash_node, void *data)
{
  t_broker_cbdef *broker_cbdef = (t_broker_cbdef *) hash_node;

  while (broker_cbdef)
  {
    t_broker_cbdef *next = broker_cbdef->next;

    StringArrayFree (broker_cbdef->proto);
    if(HASH_TABLE_NODE_KEY(&(broker_cbdef->node)))
      Free (HASH_TABLE_NODE_KEY(&(broker_cbdef->node)));
    Free (broker_cbdef);
    broker_cbdef = next;
  }
}

void
DiabloBrokerFini ()
{
  if (brokerht)
  {
    HashTableWalk (brokerht, BrokerCbdefFree, NULL);
    HashTableFree (brokerht);
  }
}

/* vim: set shiftwidth=2 expandtab cinoptions=p5,t0,(0, foldmethod=marker tw=80 cindent: */
