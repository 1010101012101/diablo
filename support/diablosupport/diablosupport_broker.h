#include <diablosupport.h>
#ifdef DIABLOSUPPORT_FUNCTIONS
#ifndef DIABLOSUPPORT_BROKER_FUNCTIONS
#define DIABLOSUPPORT_BROKER_FUNCTIONS
t_bool DiabloBrokerCallExists (const char *name);
void DiabloBrokerCallInstall (const char *name, const char *prototype, void *implementation, t_bool final, ...);
t_bool DiabloBrokerCall (const char *name, ...);
void DiabloBrokerFini ();
#endif
#endif
/* vim: set shiftwidth=2 expandtab cinoptions=p5,t0,(0, foldmethod=marker tw=80 cindent: */
