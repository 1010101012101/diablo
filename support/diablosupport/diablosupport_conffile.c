/* This research is supported by the European Union Seventh Framework Programme (FP7/2007-2013), project ASPIRE (Advanced  Software Protection: Integration, Research, and Exploitation), under grant agreement no. 609734; on-line at https://aspire-fp7.eu/. */

#include <diablosupport.h>

t_hash_table * conf_hash;
extern FILE* ConfParserin;

int ConfParserparse(void);

typedef struct 
{
  t_hash_table_node node;
  t_string  string;
} t_conf_value;

void 
ConfValueFree(const void * tf, void *unused_data)
{
  const t_conf_value* cv=tf;
  Free(cv->string);
  Free(HASH_TABLE_NODE_KEY(&cv->node));
  Free(cv);
}

t_conf_value * 
ConfValueNew(t_string x, t_string y)
{
  t_conf_value * ret=Malloc(sizeof(t_conf_value));
  HASH_TABLE_NODE_SET_KEY(&ret->node, x);
  ret->string=y;
  return ret;
}

void
ConfValueSet(t_string x, t_string y)
{
  HashTableInsert(conf_hash,ConfValueNew(x,y));
}

t_string 
ConfValueGet(t_hash_table * in,t_string x)
{
  t_conf_value * y;
  if(!in) return NULL;
  y=((t_conf_value *) HashTableLookup(in,x));
  if (y) return y->string;
  return NULL;
}

t_hash_table * 
ConfFileRead(t_const_string conffile)
{
  FILE * f = fopen(conffile,"r");
  if (!f) return NULL;
  ConfParserin = f;
  conf_hash=HashTableNew(7,0,(t_hash_func)StringHash, (t_hash_cmp)StringCmp,ConfValueFree);
  ConfParserparse();
  return conf_hash;	
}

void 
ConfValuePrint(const void * string1, void * string2)
{
  const t_conf_value * value = (const t_conf_value*)string1;
  VERBOSE(0,("(%s,%s)",HASH_TABLE_NODE_KEY(&value->node),value->string));
}
