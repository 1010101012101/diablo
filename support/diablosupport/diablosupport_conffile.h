/* This research is supported by the European Union Seventh Framework Programme (FP7/2007-2013), project ASPIRE (Advanced  Software Protection: Integration, Research, and Exploitation), under grant agreement no. 609734; on-line at https://aspire-fp7.eu/. */

#include <diablosupport.h>

#ifdef DIABLOSUPPORT_FUNCTIONS
#ifndef DIABLOSUPPORT_CONFFILE_FUNCTIONS
#define DIABLOSUPPORT_CONFFILE_FUNCTIONS
void ConfValueFree(const void * tf, void *unused_data);
t_string ConfValueGet(t_hash_table * in,t_string x);
t_hash_table * ConfFileRead(t_const_string conffile);
void ConfValuePrint(const void * string1, void * string2);
void ConfValueSet(t_string x, t_string y);
#endif
#endif
