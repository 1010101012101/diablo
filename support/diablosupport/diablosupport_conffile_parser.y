%{
#include <diablosupport.h>
#include "diablosupport_conffile_parser_extra.h"
%}
/* The parser for linker descriptions. See kernel/diablo_linkers.c for more
 * info on linker descriptions */

/* Token definitions */
%token VARIABLE
%token OPER_IS

/* Datatypes used in the parser */
%union 
{
   char * string;
}

/* Type rules */

%type <string> VARIABLE
%{
/*#include <diablosupport.h>*/
extern int confline;
extern int yylex (void);
void yyerror(const char * in)
{
  printf("At line %d: Parse error: %s\n",confline,in);
}
%}
%%

conffile: /* Empty */
	| conffile assignment {}
	;

assignment: VARIABLE OPER_IS VARIABLE { ConfValueSet($1,$3);  }
	  ;

%%

