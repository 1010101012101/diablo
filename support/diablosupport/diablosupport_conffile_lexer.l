%option prefix="ConfParser"
%option nounput
%{
#include <diablosupport.h>
#include <diablosupport_conffile_parser_extra.h>
#include <diablosupport_conffile_parser.h>

extern t_hash_table * conf_hash;
t_string long_string;
int confline=1;

void conferror(char * mesg)
{
  FATAL(("Lexer error at line %d in file: %s",confline,mesg));
}



%}
%x COMMENT
%x STRING
%%

"#"			{ BEGIN(COMMENT); }
<COMMENT>"\n"		{ BEGIN(INITIAL); confline++; }
<COMMENT>.		{ }
"\""			{  long_string=StringDup(""); BEGIN(STRING); }
<STRING>[^\"\n\\]*	{ t_string tmp=long_string; long_string=StringConcat2(tmp,yytext); Free(tmp); }
<STRING>"\\\""		{ t_string tmp=long_string; long_string=StringConcat2(tmp,"\""); Free(tmp); }
<STRING>"\\"		{ t_string tmp=long_string; long_string=StringConcat2(tmp,"\\"); Free(tmp); }
<STRING>"\""		{ ConfParserlval.string=long_string;BEGIN(INITIAL); return VARIABLE; }
<STRING>"\n"		{ conferror("Newline in string variable"); }
" "			{ } 
[^\ "=\n#][^\ "=\n]*	{ ConfParserlval.string=StringDup(yytext); return VARIABLE; }
"="			{ return OPER_IS; }
"\n"			{ confline++; }

%%
int yywrap()
{
  return 1;
}
