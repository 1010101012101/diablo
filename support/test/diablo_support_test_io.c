#include <diablosupport.h>

t_string XModifier(t_string modifiers, va_list * ap)
{
   return StringConcat3("<replaced by the modifier x (arg = ",va_arg(*ap,char *),")>");
}

void StatusIo(t_uint32 id, t_string out)
{
   printf(out);
}

int main()
{
   IoHandlerAdd(E_DEBUG,StatusIo);
   IoHandlerAdd(E_FATAL,StatusIo);
   IoModifierAdd('@','x',"bc",XModifier);
   IoModifierAdd('%',0,NULL,IoModifierPrintf);
   DEBUG(("Hallo %s\n","String"));
   DEBUG(("Hallo %d\n",10));
   DEBUG(("Hallo %d %d\n",10,20));
   DEBUG(("Hallo %d %0.2lf %s\n",2,0.3,"BRRRRR"));
   DEBUG(("Long Long 1: %lld, string: %s\n",1LL,"BRRRRR"));
   DEBUG(("Hallo @b\n"));
   DEBUG(("Hallo @@\n"));
   DEBUG(("Hallo @bx @cx\n","a","b"));
   DEBUG(("Hallo @cx @bx\n","a","b"));
   FATAL(("FATAL: Hallo @cx @bx\n","a","b"));
   return 0;
}
