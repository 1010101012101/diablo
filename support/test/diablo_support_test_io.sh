#/bin/bash

#if test -z "$srcdir"; then
#    srcdir=echo "$0" | sed s,[^/]*$,,
#    test "$srcdir" = "$0" && srcdir=.
#    test -z "$srcdir" && srcdir=.
#    test "${VERBOSE+set}" != set && VERBOSE=1
#fi






./diablo_support_test_io > tmp_out
if [ $? -eq 255 ] 
then
diff tmp_out $srcdir/data/diablo_support_test_io.out > /dev/null 2>/dev/null
RET=$?
rm tmp_out
exit $RET
else
exit 1
rm tmp_out
fi
