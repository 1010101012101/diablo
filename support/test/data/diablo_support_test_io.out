Hallo String
Hallo 10
Hallo 10 20
Hallo 2 0.30 BRRRRR
Long Long 1: 1, string: BRRRRR
Hallo @b
Hallo @
Hallo <replaced by the modifier x (arg = a)> <replaced by the modifier x (arg = b)>
Hallo <replaced by the modifier x (arg = a)> <replaced by the modifier x (arg = b)>
FATAL: Hallo <replaced by the modifier x (arg = a)> <replaced by the modifier x (arg = b)>
